docs/design/mono/wasm-threads.md

```diff
@@ -0,0 +1,239 @@
+11. throw PNSE any time C# code needs to block
+- It's not deterministic and you can't really test your app to prove it harmless.
+12. throw PNSE any time C# code or VM code needs needs to block
+- Mono VM needs to hold lock while allocating memory, even on the UI thread.
+13. modify `ConfigureAwait()`, work queue etc, to never dispatch to another thread.
+- this would probably break user code expectations about dynamic behavior of tasks and how they run in parallel with each other.
+
+# Design proposal details
+
+## UI thread
+- this is the main browser "thread", the one with DOM on it
+- will start emscripten as usual
+    - this includes C# which runs during mono startup
+- will create Deputy worker as C# thread
+- dispatch execution of C# `Task Main()` to the deputy worker.
+- dispatch all async JSImport calls to the deputy worker.
+- dispatch all async callbacks to the deputy worker.
+- throw PNSE on any JSImport call from JS
+- it will be valid C# thread, but not used directly by user code.
+    - we will try to prevent user code from running on it and from needing to do so
+- we will spin lock only for Mono VM code
+    - we assume that Mono VM will block only shortly for operations like:
+        - alloc/free memory
+        - transform IL -> IR and update Mono VM shared state
+    - we will spin lock before Blazor `renderBatch`
+        - to wait for "pause GC"
+    - we will spin lock during GC, if we hit barrier
+        - TODO: is that short enough ?
+    - we should never block for file operations or for network operations
+        - TODO: how to prove it ?
+        - Could we unregister the thread from Mono VM ? No, because we need the C# to dispatch calls in both directions in
+- it will actually execute small chunks of C# code
+    - the pieces which are in the generated code of the JSExport
+    - containing dispatch to deputy worker's synchronization context
+- could sync void C# methods be dispatched as fire and forget ?
+    - no, because that would break the contract that they are blocking until finished.
+    - also the errors would not propagate
+- TODO: is there anything special about error propagation over the interop/thread boundary ?
+
+## Deputy worker
+- this is new concept introduced here. I needed some name for it 🤷‍♂️
+- executing all **user C# code** on behalf of the UI JavaScript "thread"
+    - that is also C# entry point `Task Main()` or `void Main()`
+    - `void Main()` would return promise that never resolves to UI JavaScript
+- this thread **could block** on synchronization primitives just fine!
+- doesn't expose JavaScript state to user code.
+    - because JSHandle has thread affinity and it's unique per JS thread.
+    - as optimization we could consider running HTTP and WS client here, instead of UI thread. But JSHandle problem.
+- has SynchronizationContext installed on it
+    - So that C# calls could be dispatched to it by runtime
+- throw PNSE on attempt to marshal sync C# delegate to UI JavaScript
```

Comment:
We could fail bit later, when you actually try to call it. But you have to call it on the right thread, because of JS affinity.
I will update the doc about it.